scripts: compliance: add support for YAMLLint

fabiobaltieri opened the pull request:
Hi! Here's a proposal to integrate yamllint to our compliance tests. This is using YAMLLint (https://yamllint.readthedocs.io/en/stable/index.html), which is also used in Linux (https://lxr.missinglinkelectronics.com/linux/Documentation/devicetree/bindings/Makefile).

Linux config: https://lxr.missinglinkelectronics.com/linux/Documentation/devicetree/bindings/.yamllint
Rules documentation: https://yamllint.readthedocs.io/en/stable/rules.html

Current rules: (updated)
```
{'braces': {'forbid': False,
            'level': 'error',
            'max-spaces-inside': 0,
            'max-spaces-inside-empty': -1,
            'min-spaces-inside': 0,
            'min-spaces-inside-empty': -1},
 'brackets': {'forbid': False,
              'level': 'error',
              'max-spaces-inside': 0,
              'max-spaces-inside-empty': -1,
              'min-spaces-inside': 0,
              'min-spaces-inside-empty': -1},
 'colons': {'level': 'error', 'max-spaces-after': 1, 'max-spaces-before': 0},
 'commas': {'level': 'error',
            'max-spaces-after': 1,
            'max-spaces-before': 0,
            'min-spaces-after': 1},
 'comments': {'ignore-shebangs': True,
              'level': 'warning',
              'min-spaces-from-content': 1,
              'require-starting-space': True},
 'comments-indentation': {'level': 'warning'},
 'document-end': False,
 'document-start': {'level': 'warning', 'present': False},
 'empty-lines': {'level': 'error', 'max': 2, 'max-end': 0, 'max-start': 0},
 'empty-values': False,
 'float-values': False,
 'hyphens': {'level': 'error', 'max-spaces-after': 1},
 'indentation': {'check-multi-line-strings': False,
                 'indent-sequences': 'consistent',
                 'level': 'error',
                 'spaces': 2},
 'key-duplicates': {'level': 'error'},
 'key-ordering': False,
 'line-length': {'allow-non-breakable-inline-mappings': False,
                 'allow-non-breakable-words': True,
                 'level': 'error',
                 'max': 100},
 'new-line-at-end-of-file': {'level': 'error'},
 'new-lines': {'level': 'error', 'type': 'unix'},
 'octal-values': False,
 'quoted-strings': False,
 'trailing-spaces': {'level': 'error'},
 'truthy': {'allowed-values': ['true', 'false'],
            'check-keys': False,
            'level': 'warning'}}
```

...and I added some tweaks for the github workflow files to avoid breaking long command lines and to allow on/off booleans.

Now this is finding a lot of lint warnings in our current files:

```
$ yamllint -f parsable -c .yamllint $( find -regex '.*\.y[a]*ml' )  | sed -r 's/^.*\((.*)\)$/\1/' | sort | uniq -c
     22 brackets
     16 colons
      1 commas
    178 comments
     31 comments-indentation
      1 empty-lines
     47 hyphens
   1763 indentation
     14 key-duplicates
    192 line-length
     48 truthy
```

Full list: https://gist.github.com/fabiobaltieri/c5b72fe03b2e0ca57b3a5df922666549

The ones I checked look reasonable to me, and would have to be addressed before submitting this PR otherwise they'd block unrelated patches (case in point, I had to fix a True -> true for this one). I'll work on those on separate PRs if we decide to go ahead with this, will likely tweak few rules in the process I imagine.

One cool thing of this beside the file format consistency is that it catches duplicate keys (the `14 key-duplicates` above), which are hard to spot otherwise.

Prerequisite PRs:

- [x] Fix most errors https://github.com/zephyrproject-rtos/zephyr/pull/53449
- [x] Fix duplicate keys (https://github.com/zephyrproject-rtos/zephyr/pull/53462)
- [x] Fix indentations (https://github.com/zephyrproject-rtos/zephyr/pull/53481)

fabiobaltieri committed:
scripts: compliance: add support for YAMLLint

Add a YAMLLint compliance check that uses the yamllint package to report
linting error on YAML files.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

gmarull commented on scripts/ci/check_compliance.py:
```diff
@@ -991,6 +993,34 @@ def run(self):
+
+            yaml_config = config.YamlLintConfig(file=config_file)
+
+            if file.startswith(".github/"):
+                # Tweak few rules for workflow files.
+                yaml_config.rules["line-length"] = False
+                yaml_config.rules["truthy"]["allowed-values"].extend(['on', 'off'])
```
should we have multiple `.yamllint` files in relevant subdirs? e.g. one for workflows, one for bindings?

fabiobaltieri replied:
Yeah I thought a good bit about how to manage overrides. I'm thinking: if we can get away with only this override we may as well just keep it like this, it's nice and compact. If we start having more per directory configuration though we may as well have separate configs.

fabiobaltieri replied:
Well, I added another one, but it's at the root directory so per subdir config override does not work there. I'm open to ideas.

gmarull commented:
+1, would be nice to have an autoformatter that complies with such settings as well

stephanosio reviewed (commented):
This looks reasonable, thanks. It has been in the CI internal issue tracker for a while -- great to see it finally being addressed.

fabiobaltieri commented:
> This looks reasonable, thanks. It has been in the CI internal issue tracker for a while -- great to see it finally being addressed.

Cool, good to know, I'll start sorting out the existing compliance errors then, think I"ll do two or three different separate PRs, then we can move forward with this.

fabiobaltieri commented:
Sent a first round of fixes in https://github.com/zephyrproject-rtos/zephyr/pull/53449, changed here:
- dropped the override for workflows files to allow `on`, `off`, should be safe to use `true` and `false`
- added `comments: {min-spaces-from-content: 1}` (seems reasonable, Linux has it too)

fabiobaltieri commented:
Rebased, at this stage:

```
$ yamllint -f parsable -c .yamllint $( find -regex '.*\.y[a]*ml' )  | sed -r 's/^.*\((.*)\)$/\1/' | sort | uniq -c
     33 line-length
     12 truthy
```

All in `.github` and `./.codecov.yml` with a matching override in the script.

Good to go! Thanks for keeping up with the monster PRs.

carlescufi reviewed (approved):
Nice! Thanks!
